contracts/receivables/RealWorldReceivable.sol

```diff
@@ -0,0 +1,198 @@
+    mapping(uint256 => ReceivableInfo) public receivableInfoMapping;
+
+    // The status of the receivable.
+    enum Status {
+        Unpaid,
+        Paid,
+        PartiallyPaid
+    }
+
+    /**
+     * @dev Emitted when the owner of a receivable calls the declarePayment function
+     * @param from The address of the owner of the receivable
+     * @param to The address of the Pool that's paid by the receivable
+     * @param tokenId The ID of the receivable token
+     * @param amount The amount that was declared paid
+     */
+    event PaymentDeclared(
+        address indexed from,
+        address indexed to,
+        uint256 indexed tokenId,
+        uint256 amount
+    );
+
+    /// @custom:oz-upgrades-unsafe-allow constructor
+    constructor() {
+        _disableInitializers();
+    }
+
+    /**
+     * @dev Initializer that sets the default admin and minter roles
+     */
+    function initialize() public initializer {
+        __ERC721_init("RealWorldReceivable", "RWR");
+        __ERC721Enumerable_init();
+        __ERC721URIStorage_init();
+        __ERC721Burnable_init();
+        __AccessControl_init();
+
+        _grantRole(DEFAULT_ADMIN_ROLE, msg.sender);
+        _grantRole(MINTER_ROLE, msg.sender);
+    }
+
+    /**
+     * @dev Creates a new receivable token and assigns it to the recipient address
+     * @param recipient The address that will receive the new token
+     * @param paymentAddress The address that's expected to be paid for this receivable
+     * @param paymentToken The ERC20 token used to pay the receivable
+     * @param receivableAmount The total amount of the receivable
+     * @param maturityDate The date at which the receivable becomes due
+     * @param uri The URI of the metadata associated with the receivable
+     */
+    function safeMint(
+        address recipient,
+        address paymentAddress,
+        address paymentToken,
+        uint96 receivableAmount,
+        uint64 maturityDate,
+        string memory uri
+    ) public onlyRole(MINTER_ROLE) {
+        uint256 tokenId = _tokenIdCounter.current();
+        _tokenIdCounter.increment();
+        _safeMint(recipient, tokenId);
+
+        ReceivableInfo memory receivableInfo = ReceivableInfo(
+            paymentAddress,
+            paymentToken,
+            receivableAmount,
+            0, // paidAmount
+            maturityDate
+        );
+        receivableInfoMapping[tokenId] = receivableInfo;
+
+        _setTokenURI(tokenId, uri);
+    }
+
+    /**
+     * @dev Declares payment for a receivable.
+     * Only the owner of the token can declare a payment.
+     * The payment method for the receivable must be Declarative.
+     * The receivable must not already be paid in full.
+     * Emits a `PaymentDeclared` event.
+     * @param tokenId The ID of the receivable token.
+     * @param paymentAmount The amount of payment being declared.
+     */
+    function declarePayment(uint256 tokenId, uint96 paymentAmount) public {
```

Comment:
Since this is only declaring payments, no ERC20 tokens are actually sent in this function. It would be an unused variable, so I decided it's not useful to include here.

Reply:
You are using it when a receivable is created though. We should be consistent for creation and declaring payments. Honestly, the value for fields paymentToken and currencyCode is limited but they add quite some cognitive load. I feel the combination of paymentToken and currencyCode is the source of complexity. Since this is RWR, do we need paymenToken at all?

Reply:
We don't necessarily need to have it, I just included it to be flexible. In creation a Receivable enforces that it only has one of either paymentToken or currencyCode, so it can support representing both fiat and crypto income. If we don't think many companies will have crypto income though I can remove this.